core/services/ocr2/plugins/threshold/decryption_queue.go

```diff
@@ -0,0 +1,208 @@
+	if len(ciphertext) > int(dq.maxCiphertextBytes) {
+		return nil, errors.New("ciphertext too large")
+	}
+
+	pendingChan, err := dq.getResult(ciphertextId, ciphertext)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case pt := <-pendingChan:
+		return pt, nil
+	case <-ctx.Done():
+		dq.mu.Lock()
+		defer dq.mu.Unlock()
+		delete(dq.pendingRequests, string(ciphertextId))
+		return nil, errors.New("cancelled")
+	}
+}
+
+func (dq *decryptionQueue) getResult(ciphertextId CiphertextId, ciphertext []byte) (chan []byte, error) {
+
+	dq.mu.Lock()
+	defer dq.mu.Unlock()
+
+	pendingChan := make(chan []byte, 1)
+
+	req, ok := dq.completedRequests[string(ciphertextId)]
+	if ok {
+		pendingChan <- req.plaintext
+		req.timer.Stop()
+		delete(dq.completedRequests, string(ciphertextId))
+		return pendingChan, nil
+	}
+
+	_, isDuplicateId := dq.pendingRequests[string(ciphertextId)]
+	if isDuplicateId {
+		return nil, errors.New("Decrypt ciphertextId must be unique")
+	}
+
+	if uint32(len(dq.pendingRequestQueue)) >= dq.maxQueueLength {
+		return nil, errors.New("queue is full")
+	}
+	dq.pendingRequestQueue = append(dq.pendingRequestQueue, ciphertextId)
+
+	dq.pendingRequests[string(ciphertextId)] = pendingRequest{
+		pendingChan,
+		ciphertext,
+		uint64(time.Now().Unix()) + dq.completedRequestsCacheTimeoutMs/1000,
+	}
+
+	return pendingChan, nil
+}
+
+func (dq *decryptionQueue) GetRequests(requestCountLimit uint32, totalBytesLimit uint32) []DecryptionRequest {
+	dq.mu.Lock()
+	defer dq.mu.Unlock()
+
+	requests := make([]DecryptionRequest, 0, requestCountLimit)
+	totalBytes := 0
+
+	for i := 0; len(requests) < int(requestCountLimit); i++ {
+		if i >= len(dq.pendingRequestQueue) {
+			break
+		}
+
+		requestId := dq.pendingRequestQueue[i]
+		pendingRequest, exists := dq.pendingRequests[string(requestId)]
+
+		if !exists {
+			dq.pendingRequestQueue = append(dq.pendingRequestQueue[:i], dq.pendingRequestQueue[i+1:]...)
+			i--
+			continue
+		}
+
+		requestToAdd := DecryptionRequest{
+			requestId,
+			pendingRequest.ciphertext,
+		}
+
+		additionalBytes := len(requestId) + len(pendingRequest.ciphertext)
+
+		if (totalBytes + additionalBytes) > int(totalBytesLimit) {
+			continue
+		}
+
+		requests = append(requests, requestToAdd)
+		totalBytes += additionalBytes
+	}
+
+	return requests
+}
+
+func (dq *decryptionQueue) GetCiphertext(ciphertextId CiphertextId) ([]byte, error) {
+	dq.mu.Lock()
+	defer dq.mu.Unlock()
+
+	req, ok := dq.pendingRequests[string(ciphertextId)]
+	if !ok {
+		return nil, errors.New("ciphertext not found")
+	}
+
+	return req.ciphertext, nil
+}
+
+func (dq *decryptionQueue) ResultReady(ciphertextId CiphertextId, plaintext []byte) {
+	dq.mu.Lock()
+	defer dq.mu.Unlock()
+
+	req, ok := dq.pendingRequests[string(ciphertextId)]
+	if ok {
+		req.pendingChan <- plaintext
+		delete(dq.pendingRequests, string(ciphertextId))
+	} else {
```

Comment:
Add a comment for this block along the lines of "Cache ready result in completedRequests map for cacheTimeoutMs to account for delayed Decrypt() calls."